pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1106,12 +1182,356 @@ func (s *sqlEntityServer) List(ctx context.Context, r *entity.EntityListRequest)
+				oneExtra: true, // request one more than the limit (and show next token if it exists)
+				orderBy:  []string{"resource_version"},
+			}
+
+			entityQuery.addWhere("resource_version > ?", since)
+
+			query, args := entityQuery.toQuery()
+
+			rows, err := s.sess.Query(ctx, query, args...)
+			if err != nil {
+				return err
+			}
+			defer func() { _ = rows.Close() }()
+
+			found := int64(0)
+			for rows.Next() {
+				found++
+				if found > entityQuery.limit {
+					return nil
+				}
+
+				result, err := s.rowToEntity(ctx, rows, rr)
+				if err != nil {
+					return err
+				}
+
+				if result.ResourceVersion > since {
+					since = result.ResourceVersion
+				}
+
+				s.log.Debug("sending poll result", "guid", result.Guid, "action", result.Action, "rv", result.ResourceVersion)
+				out <- result
+			}
+
+			hasmore = false
+			return nil
+		}()
+		if err != nil {
+			return since, err
+		}
+	}
+
+	return since, nil
+}
+
+func watchMatches(r *entity.EntityWatchRequest, result *entity.Entity) bool {
+	// Resource version too old
+	if result.ResourceVersion <= r.Since {
+		return false
+	}
+
+	// Folder guid
+	if r.Folder != "" && r.Folder != result.Folder {
+		return false
+	}
+
+	// must match at least one resource if specified
+	if len(r.Resource) > 0 {
+		matched := false
+		for _, res := range r.Resource {
+			if res == result.Resource {
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			return false
+		}
+	}
+
+	// must match at least one key if specified
+	if len(r.Key) > 0 {
+		matched := false
+		for _, k := range r.Key {
+			key, err := entity.ParseKey(k)
+			if err != nil {
+				return false
+			}
+
+			if key.Namespace == result.Namespace && key.Group == result.Group && key.Resource == result.Resource && (key.Name == "" || key.Name == result.Name) {
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			return false
+		}
+	}
+
+	// must match at least one label/value pair if specified
+	// TODO should this require matching all label conditions?
+	if len(r.Labels) > 0 {
+		matched := false
+		for labelKey, labelValue := range r.Labels {
+			if result.Labels[labelKey] == labelValue {
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			return false
+		}
+	}
+
+	return true
+}
+
+// watch is a helper to get the next set of entities and send them to the client
+func (s *sqlEntityServer) watch(ctx context.Context, r *entity.EntityWatchRequest, w entity.EntityStore_WatchServer) error {
+	s.log.Debug("watch started", "since", r.Since)
+
+	evts, err := s.broadcaster.Subscribe(w.Context())
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		// user closed the connection
+		case <-w.Context().Done():
+			return nil
+		// got a raw result from the broadcaster
+		case result := <-evts:
+			// result doesn't match our watch params, skip it
+			if !watchMatches(r, result) {
+				s.log.Debug("watch result not matched", "guid", result.Guid, "action", result.Action, "rv", result.ResourceVersion)
+				break
```

Comment:
If I understand the code correctly we should continue on to the next event instead of stopping the watch, shouldn't we?
```suggestion
				continue
```

Reply:
more or less the same impact since we go to the next loop of the for and call select again, but yes continue may make more sense